RotatedSortedArray.py

```diff
@@ -0,0 +1,99 @@
+# Time Complexity : O(log(n))
+# Space Complexity : O(1)
+# Did this code successfully run on Leetcode : yes
+# Any problem you faced while coding this : No
+
+
+# Iterative Solution
+class Solution:
+
+    def search(self, nums: List[int], target: int) -> int:
+
+        #Intializing the left and right values which is start and end, In python we can give this by 0(start) and len(nums)-1 (end)
+        left = 0
+        right = len(nums) - 1
+
+        #If this case fails means we reached end of the list and no numbers to search and we will return -1
+        while left <= right:
+            #Calculating mid => this way of calculting mid will prevent overflow.
+            mid = left + (right - left) // 2
+
+            #If mid is the target we will return it
+            if nums[mid] == target:
+                return mid
+
+            #If this case pass - means it is left sorted else it is right sorted
+            if nums[left] <= nums[mid]:
+
+                # If this case pass - means the target is between this left and mid and we convert it to new range by keeping the left same and right to mid -1
+                #Else vice versa of it
+                if nums[left] <= target <= nums[mid]:
+
+                    right = mid - 1
+
+                else:
+
+                    left = mid + 1
+            else:
+
+                # If this case pass - means the target is between this Right and mid and we convert it to new range by keeping the Right same and left to mid +1
+                # Else vice versa of it
+                if nums[right] >= target >= nums[mid]:
+
+                    left = mid + 1
+                else:
+
+                    right = mid - 1
+
+        return -1
+
+
+# Recursive
+
+# Time Complexity : O(log(n))
+# Space Complexity : O(n) - I have no clear explanation about this. This is just a random guess.please give me the explanation for this. Thanks.
```

Comment:
When we use a recursive approach, then at each stage, we make a recursive call. That means leaving the current invocation on the stack and calling a new one. When we are K levels deep, we have got k lots of stack frame, so the space complexity ends up being proportional to the depth you have to search. Space Complexity O(log n).